pkg/trace/stats/sublayers.go

```diff
@@ -34,6 +39,195 @@ func (v SublayerValue) GoString() string {
+func (a *activeSpanSet) remove(spanIdx int) {
+	i := a.indexes[spanIdx]
+	if i != -1 {
+		a.indexes[a.spans[a.size-1]] = i
+		a.spans[i] = a.spans[a.size-1]
+		a.size--
+		a.indexes[spanIdx] = -1
+	}
+
+}
+
+// clear clears the set. It prepares it to hold maximum n elements
+func (a *activeSpanSet) clear(n int) {
+	a.size = 0
+	for i := 0; i < n; i++ {
+		a.indexes[i] = -1
+		a.spans[i] = 0
+	}
+}
+
+// resize updates the capacity of the set
+func (a *activeSpanSet) resize(capacity int) {
+	a.capacity = capacity
+	a.spans = make([]int, capacity)
+	a.indexes = make([]int, capacity)
+}
+
+// SublayerCalculator holds arrays used to compute sublayer metrics.
+// Re-using arrays reduces the number of allocations
+// A sublayer metric is the execution duration of a given type / service in a trace
+// The metrics generated are detailed here: https://docs.datadoghq.com/tracing/guide/metrics_namespace/#duration-by
+// The arrays in this structure are updated as the computeExecDurations traverses all the sorted timestamps of a trace.
+// They are indexed by the span index in the trace (for eg, openSpans[0] == true means that the first span of the trace is open)
+type SublayerCalculator struct {
+	// openSpans holds whether each span is open. A span is open if it has started, but hasn't ended yet.
+	openSpans []bool
+	// nChildren is the number of direct children spans that are currently open
+	nChildren []int
+	// activeSpans is a set active spans. A span is active at a given timestamp, if it is open and if it has no open children.
+	activeSpans *activeSpanSet
+	// Exec duration: for each period where the span is active, its Exec Duration increases by periodDuration/numberOfActiveSpans
+	//     so if a span is active during 20ms, in parallel with another span active at the same time, its exec duration will be 10ms
+	execDuration []float64
+	// timestamps are the sorted timestamps (starts and ends of each span) of a trace
+	timestamps sortableTimestamps
+	// capacity specifies the maximum trace size in spans the calculator can process without re-allocations
+	capacity int
+}
+
+// NewSublayerCalculator returns a new SublayerCalculator.
+func NewSublayerCalculator() *SublayerCalculator {
+	c := &SublayerCalculator{activeSpans: &activeSpanSet{}}
+	c.resize(defaultCalculatorSpanCapacity)
+	return c
+}
+
+// resize allocates arrays of the sublayer calculator
+// it should be called every time we receive a trace with more spans than the capacity of the calculator
+func (c *SublayerCalculator) resize(capacity int) {
+	c.capacity = capacity
+	c.openSpans = make([]bool, capacity)
+	c.nChildren = make([]int, capacity)
+	c.execDuration = make([]float64, capacity)
+	c.timestamps = make(sortableTimestamps, 2*capacity)
+	c.activeSpans.resize(capacity)
+}
+
+// clear clears structures of the sublayer calculator to prepare the computation of sublayer metrics
+// for a trace with max n spans
+func (c *SublayerCalculator) clear(n int) {
```

Comment:
Can we eliminate this method and just put it at the end of `resize`? It isn't really re-used. I found the name `reset` would be better when merging the two.